### PR TITLE
Correlation Improvements

### PR DESCRIPTION
**Pull request by jcdickinson**
- Change Header option to a List, allowing multiple choices.
- Allow null correlation header in context (to signal no correlation ID header).

### CODE REVIEW

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationContext.cs [hunk 1]
```diff
@@ -12,9 +12,8 @@ internal CorrelationContext(string correlationId, string header)
         {
             if (string.IsNullOrWhiteSpace(correlationId))
                 throw new ArgumentException(ArgumentNullOrWhitespace, nameof(correlationId));
-
-            if (string.IsNullOrWhiteSpace(header))
-                throw new ArgumentException(ArgumentNullOrWhitespace, nameof(header));
+            if (header != null && string.IsNullOrWhiteSpace(header))
+                throw new ArgumentOutOfRangeException(ArgumentNullOrWhitespace, nameof(CorrelationId));
```
Comment (pbbadenhorst) [inline src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationContext.cs]: Undo this change please

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationHeader.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            return Name == other.Name &&
+                   IncludeInResponse == other.IncludeInResponse;
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            var hashCode = -406407647;
+            hashCode = hashCode * -1521134295 + StringComparer.Ordinal.GetHashCode(Name);
+            hashCode = hashCode * -1521134295 + IncludeInResponse.GetHashCode();
```
Comment (pbbadenhorst) [inline src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationHeader.cs]: Should this not be a second different prime

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationMiddleware.cs [hunk 3]
```diff
@@ -80,34 +80,45 @@ public Task Invoke(HttpContext context)
...
-            if (headerExists && !StringValues.IsNullOrEmpty(headerValue))
-                return headerValue;
+            for (var i = 0; i < Options.Headers.Count; i++)
+            {
+                CorrelationHeader header = Options.Headers[i];
+                var headerExists = context.Request.Headers.TryGetValue(header.Name, out StringValues headerValue);
+                if (headerExists && !StringValues.IsNullOrEmpty(headerValue))
+                {
+                    headerName = header.Name;
+                    return headerValue;
```
Comment (pbbadenhorst) [inline src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationMiddleware.cs]: Should we not be handling all the specified headers?
I.e. if the options contains three correlation headers we should look at all three created values for each missing one and then add each in turn to the response as needed?

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationOptions.cs [hunk 4]
```diff
@@ -13,29 +14,24 @@ public class CorrelationOptions
...
+        public const string DefaultHeader = "X-Correlation-ID";
 
         /// <summary>
-        /// Gets or sets whether the <see cref="HttpContext.TraceIdentifier"/> of the context should be used if not correlation id
+        /// Gets or sets whether the <see cref="HttpContext.TraceIdentifier"/> of the context should be used if no correlation ID
         /// is present in request (defaults to false).
         /// If <c>false</c>, <see cref="CorrelationIdGenerator"/> is invoked to generate the correlation id.
         /// </summary>
-        public bool UseTraceIdentifier { get; internal set; } = false;
+        public bool UseTraceIdentifier { get; set; } = false;
```
Comment (pbbadenhorst) [inline src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationOptions.cs]: This should move to the CorrelationHeader struct so each header can be handled on it's own

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationHeader.cs (hunk 2 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationHeader.cs]: This is autogen code. I can't imagine this structure being used in a hash structure anyway.

File: src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationContext.cs (hunk 1 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationContext.cs]: What should the value be if no header is found?

Comment (jcdickinson) [reply src/SourceCode.Clay.AspNetCore.Middleware.Correlation/CorrelationContext.cs]: Ignore; I have a plan
